### PR TITLE
Backup and restore to report errors found from segments' status file

### PR DESCRIPTION
**Pull request by tangp3**
Gpcrondump and gpdbrestore redirect stderr into status file for segments,
errors should be reflected into the master's report file.
Warnings will be given if process completed but errors were generated in
the report file.

Also made some refactor of the code.

Authors: Pengcheng Tang, Nikhil Kak

### CODE REVIEW

File: gpMgmt/bin/gppylib/operations/restore.py [hunk 1]
```diff
@@ -439,8 +439,8 @@ def validate_tablenames(table_list):
 
 class RestoreDatabase(Operation):
     def __init__(self, restore_timestamp, no_analyze, drop_db, restore_global, master_datadir, backup_dir,
-                 master_port, dump_dir, dump_prefix, no_plan, restore_tables, batch_default, no_ao_stats,
-                 redirected_restore_db, report_status_dir, restore_stats, metadata_only, ddboost,
+                 master_port, dump_dir, dump_prefix, no_plan, restore_tables, batch_default,
+                 no_ao_stats, redirected_restore_db, report_status_dir, restore_stats, metadata_only, ddboost,
```
Comment (jimmyyih) [inline gpMgmt/bin/gppylib/operations/restore.py]: Why was this diff done?  It doesn't seem to do anything beneficial.

Comment (tangp3) [inline gpMgmt/bin/gppylib/operations/restore.py]: That was touched during resolving conflict during rebase, functionally no difference.

File: src/bin/pg_dump/cdb/cdb_restore.c [hunk 2]
```diff
@@ -1208,6 +1210,20 @@ threadProc(void *arg)
...
+	 * for ERRORS and report them if found
+	 */
+	if(pParm->bSuccess || pParm->pszErrorMsg == NULL)
+	{
+		pqBuffer = createPQExpBuffer();
+		int status = ReadBackendBackupFileError(pConn, status_file_dir, pszKey, BFT_RESTORE_STATUS, progname, pqBuffer);
+		if (status != 0)
+		{
+			pParm->pszErrorMsg = pqBuffer->data;
+		}
```
Comment (danielgustafsson) [inline src/bin/pg_dump/cdb/cdb_restore.c]: I would prefer copying the string in `->data` and calling destroyPQExpBuffer()` before leaving the block.

Comment (tangp3) [inline src/bin/pg_dump/cdb/cdb_restore.c]: Hi Daniel:
Thanks for reviewing.  This block of code line is right before the program exits, so I left the system to recollect the memory.

I guess making a copy eventually needs the similar memory allocation.

**Pull request merged**